Gem/Code/Source/Weapons/TraceWeapon.cpp

```diff
@@ -33,7 +33,14 @@ namespace MultiplayerSample
             if (isMultiSegmented)
             {
                 ActiveShot activeShot{ eventData.m_initialTransform, eventData.m_targetPosition, LifetimeSec{ 0.0f } };
-                weaponState.m_activeShots.emplace_back(activeShot);
+                if (weaponState.m_activeShots.size() < weaponState.m_activeShots.max_size())
+                { 
+                    weaponState.m_activeShots.emplace_back(activeShot);
+                }
+                else
+                {
+                    AZ_Assert(false, "Attempting to add too many active shots to the TraceWeapon.");
```

Comment:
Are you sure you want this to be a crash in this case? I think in some setups with asserts enabled, it can functionally feel like a crash when an assert files off. Maybe make it an AZ_Error or AZ_Warning?

Reply:
Yeah, an assert breaks under a debugger, for example.

Reply:
Yep, pretty sure I want it to be an assert, since it's the moment at which it's clear that the code is broken and it's catching a programming flaw. If other folks _really_ want it to be an AZ_Error since it's now handling the problem semi-gracefully we can do that, but that goes against our general guidance for when to use one vs the other.
Asserts - programming errors
Error - content errors
Warning - content problems that are recoverable

Reply:
You might want to re-title the pull request, then. I interpreted "Prevent crashes" as you not wanting the game state to stop when this occurs. But if you just want a controlled crash, that's fine.

Reply:
This is, or at least was, a programmer error. Although as you note, with your guard I guess it's now no longer.. I'll leave the consideration of whether or not to use an assert or an error to you. The assert is more useful if we think there might be a better fix long-term and we just need to capture an error case or two in the debugger to help determine what that may be. If we're never planning to fix this in a different manner, then the assert may be a bit heavy handed.

Reply:
@AMZN-stankowi Just for specificity, an assert isn't inherently a crash. By default, execution can continue past an assert. A user has to specifically set sys_asserts=3 (I think) to _actually_ crash from an assert. In comparison, the crash that this is intended to prevent could _never_ have execution continue, regardless of user settings. Regardless of terminology, I think we're aligned on what I'm doing here. :)

@kberg-amzn It's still a programmer error, even with the guard. I'm quite confident that if a second player fires more than 32 total shots over the course of the game, every shot past the 32nd one will not work. I don't really think the first 32 work either tbh. This 100% needs a better fix, this assert is just meant to prevent crashes in the meantime so that we can hopefully have more productive playtests.

Reply:
Yeah it's not technically a crash, but to many users, especially non-engineers, it can feel like a crash. Program execution halts, you get a popup message.

In the case of a multiplayer experience, I don't know how O3DE handles it, but previous codebases I've worked in, if you get an assert you drop connection at that point, too.

Reply:
It appears that by default, it gets written into the log, program execution does not halt, and you get no popups or onscreen messaging.

For non pak file builds, you can override the behavior for yourself in your user setreg file. For pak file builds, if you're running with a debugger, you can override the behavior by passing "-bg_assertsAutoBreak=true" so that the debugger at least breaks for you.